enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,641 @@
+that is available at cloud.redhat.com. The service is implemented with a
+traditional relational database. In order to integrate with OpenShift
+infrastructure management tooling, it is necessary to additionally expose its
+capabilities as Kubernetes-native APIs. They will be implemented as Custom
+Resource Definitions and a controller with the assisted service's capabilities.
+
+For the first deliverable, a local deployment of the backing service will
+include new controllers that will operate to achieve the desired state as
+declared in CRDs. Thus the service will expose both REST and Kubernetes-native
+APIs, and both API layers will call the same set of internal methods. The local
+service will include a database; sqlite if possible, or else postgresql.
+
+    --------------------------
+    |  REST API  |  k8s API  |
+    --------------------------
+    | Assisted Service logic |
+    --------------------------
+        |               |
+        v               v
+      SQL DB        File System
+                      storage
+
+For modest clusters that do not have persistent storage available but need
+day-2 ability to add nodes, the database will need to be reconstructable in
+case its current state is lost. Thus the CRs will be the source of truth for
+the specs, and the controller will upon statup first ensure that the DB
+reflects the desired state in CRs. The source of truth for the status, however,
+is the actual state of the agents running on the hosts being installed and not
+necessarily what was previously recorded.
+
+Agents that are running on hardware and waiting for installation-related
+instructions will continue to communicate with the assisted service using the
+existing REST API. In the future that may transition to using a CRD.
+
+**AgentDrivenInstallationImage**
+This resource represents a discovery image that should be used for booting
+hosts. In the REST API this resource is embedded in the "cluster" resource, but
+for kubernetes-native is more natural for it to be separate.
+
+Spec:
+ * clusterName: string
+ * sshPublicKey: string
+ * pullSecretName: string (reference to k8s secret)
+ * ignitionOverrides: object
+ * staticIpConfiguration: object
+
+Status:
+ * state: string
+ * sizeBytes: integer
+ * downloadUrl: string
+ * createdAt: string
+ * expiresAt: string
+
+**AgentDrivenClusterInstallation**
+This resource correlates to the "cluster" resource in the assisted installer's
+current API.
+
+Spec:
+ * name: string
+ * approved: boolean (used for starting installation)
+ * openshiftVersion: string
+ * baseDnsDomain: string
+ * clusterNetworkCidr: string
+ * clusterNetworkHostPrefix: integer
+ * serviceNetworkCidr: string
+ * apiVip: string
+ * apiVipDnsName: string
+ * ingressVip: string
+ * machineNetworkCidr: string
+ * sshPublicKey: string
+ * vipDhcpAllocation: boolean
+ * httpProxy: string
+ * httpsProxy: string
+ * noProxy: string
+ * userManagedNetworking: boolean
+ * additionalNtpSource: string
```

Comment:
This NTP source example makes me super nervous about what this resource could turn into.

AIUI, there is only one recommended way to configure the NTP source in an cluster - using a MachineConfig that lays down an /etc/chrony.conf file

If this isn't sufficiently user friendly (and I agree that it is not!) then that is a good argument for it being a top-level install-config option

A new resource like this should not accrete installation options that are "too hard" to add to the core installer

See also:

https://github.com/openshift/installer/issues/3258
https://github.com/openshift/machine-config-operator/issues/629
https://github.com/openshift/installer/pull/1951
https://github.com/openshift/machine-config-operator#applying-configuration-changes-to-the-cluster

Reply:
This field is used to synchronize clocks of hosts that have not yet formed a cluster. If hosts boot the live ISO and run the agent, and they do not already have NTP sources configured through other means (DHCP, ...), and their clocks are divergent, then we see failed installations. AIUI the user can set this value if clock validation fails, then the agent will apply it, then the validation will eventually succeed after clocks synchronize, and finally installation can be initiated.

When this field is used to add an NTP source for the above use case, a machine config is additionally created so that the new cluster will continue using that source. But this is a secondary behavior driven by the need to converge clocks pre-installation. If an install-config option were available, that could be used instead.

Here's the relevant BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1884465

Does this alleviate your concerns? I know it's still creating an API for something that could or should reasonably also be in the install config, but putting it in the install config would not solve this problem.

Reply:
Thanks, it's useful to know it serves this pre-installation purpose also

It makes the issue more complicated, but it doesn't alleviate the concern - there's nothing specific to agent-based installations about users needing to configure an NTP server. See https://github.com/openshift/hive/pull/1247#discussion_r559513984